## Types of recommendation system

1. Content based
2. Collaboration filtering
3. Hybrid

## We are building content based movie recommendation system

In [1]:
import numpy as np
import pandas as pd

In [2]:
movies = pd.read_csv('movies.csv')
credits = pd.read_csv('credits.csv')

In [3]:
movies.head()


budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [4]:
credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

## Let's merge 2 dataframes based on title

In [5]:
movies = movies.merge(credits, on='title')

In [6]:
movies

budget                                             genres  \
0     237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1     300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2     245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3     250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4     260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
...         ...                                                ...   
4804     220000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4805       9000  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
4806          0  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
4807          0                                                 []   
4808          0                [{"id": 99, "name": "Documentary"}]   

                                               homepage      id  \
0                           http://www.avatarmovie.com/   19995   
1          http://disney.go.com/disneypictures/pirates/     285   
2           http://www.sonypictures.com/movies/spectre/  206647   
3                    http://www.thedarkknightrises.com/   49026   
4                  http://movies.disney.com/john-carter   49529   
...                                                 ...     ...   
4804                                                NaN    9367   
4805                                                NaN   72766   
4806  http://www.hallmarkchannel.com/signedsealeddel...  231617   
4807                        http://shanghaicalling.com/  126186   
4808                                                NaN   25975   

                                               keywords original_language  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4     [{"id": 818, "name": "based on novel"}, {"id":...                en   
...                                                 ...               ...   
4804  [{"id": 5616, "name": "united states\u2013mexi...                es   
4805                                                 []                en   
4806  [{"id": 248, "name": "date"}, {"id": 699, "nam...                en   
4807                                                 []                en   
4808  [{"id": 1523, "name": "obsession"}, {"id": 224...                en   

                                original_title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4804                               El Mariachi   
4805                                 Newlyweds   
4806                 Signed, Sealed, Delivered   
4807                          Shanghai Calling   
4808                         My Date with Drew   

                                               overview  popularity  \
0     In the 22nd century, a paraplegic Marine is di...  150.437577   
1     Captain Barbossa, long believed to be dead, ha...  139.082615   
2     A cryptic message from Bond’s past sends him o...  107.376788   
3     Following the death of District Attorney Harve...  112.312950   
4     John Carter is a war-weary, former military ca...   43.926995   
...                                                 ...         ...   
4804  El Mariachi just wants to play his guitar and ...   14.269792   
4805  A newlywed couple's honeymoon is upended by th...    0.642552   
4806  "Signed, Sealed, Delivered" introduces a dedic...    1.444476   
4807  When ambitious New York attorney Sam is sent t...    0.857008   
4808  Ever since the second grade when he first saw ... 

In [7]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

## less important columns to create tags in recommendation system will be removed

Important columns are

1. genres
2. id ==> useful for displayimg posters in website
3. keywords  

In [8]:
movies['original_language'].value_counts()

original_language
en    4510
fr      70
es      32
zh      27
de      27
hi      19
ja      16
it      14
ko      12
cn      12
ru      11
pt       9
da       7
sv       5
nl       4
fa       4
th       3
he       3
ta       2
cs       2
ro       2
id       2
ar       2
vi       1
sl       1
ps       1
no       1
ky       1
hu       1
pl       1
af       1
nb       1
tr       1
is       1
xx       1
te       1
el       1
Name: count, dtype: int64

Since among 4809 data, 4510 data are in eng langauge so language doesn't have much stregth in recommedation

4. title
5. overview ==> if 2 movies content is similar then they are similar
6. cast ==> Actors
7. crew ==> director

In [9]:
movies = movies[['title', 'cast', 'crew', 'overview', 'genres', 'id', 'keywords']]

In [10]:
movies

title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4804                               El Mariachi   
4805                                 Newlyweds   
4806                 Signed, Sealed, Delivered   
4807                          Shanghai Calling   
4808                         My Date with Drew   

                                                   cast  \
0     [{"cast_id": 242, "character": "Jake Sully", "...   
1     [{"cast_id": 4, "character": "Captain Jack Spa...   
2     [{"cast_id": 1, "character": "James Bond", "cr...   
3     [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4     [{"cast_id": 5, "character": "John Carter", "c...   
...                                                 ...   
4804  [{"cast_id": 1, "character": "El Mariachi", "c...   
4805  [{"cast_id": 1, "character": "Buzzy", "credit_...   
4806  [{"cast_id": 8, "character": "Oliver O\u2019To...   
4807  [{"cast_id": 3, "character": "Sam", "credit_id...   
4808  [{"cast_id": 3, "character": "Herself", "credi...   

                                                   crew  \
0     [{"credit_id": "52fe48009251416c750aca23", "de...   
1     [{"credit_id": "52fe4232c3a36847f800b579", "de...   
2     [{"credit_id": "54805967c3a36829b5002c41", "de...   
3     [{"credit_id": "52fe4781c3a36847f81398c3", "de...   
4     [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   
...                                                 ...   
4804  [{"credit_id": "52fe44eec3a36847f80b280b", "de...   
4805  [{"credit_id": "52fe487dc3a368484e0fb013", "de...   
4806  [{"credit_id": "52fe4df3c3a36847f8275ecf", "de...   
4807  [{"credit_id": "52fe4ad9c3a368484e16a36b", "de...   
4808  [{"credit_id": "58ce021b9251415a390165d9", "de...   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4804  El Mariachi just wants to play his guitar and ...   
4805  A newlywed couple's honeymoon is upended by th...   
4806  "Signed, Sealed, Delivered" introduces a dedic...   
4807  When ambitious New York attorney Sam is sent t...   
4808  Ever since the second grade when he first saw ...   

                                                 genres      id  \
0     [{"id": 28, "name": "Action"}, {"id": 12, "nam...   19995   
1     [{"id": 12, "name": "Adventure"}, {"id": 14, "...     285   
2     [{"id": 28, "name": "Action"}, {"id": 12, "nam...  206647   
3     [{"id": 28, "name": "Action"}, {"id": 80, "nam...   49026   
4     [{"id": 28, "name": "Action"}, {"id": 12, "nam...   49529   
...                                                 ...     ...   
4804  [{"id": 28, "name": "Action"}, {"id": 80, "nam...    9367   
4805  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   72766   
4806  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...  231617   
4807                                                 []  126186   
4808                [{"id": 99, "name": "Documentary"}]   25975   

                                               keywords  
0     [{"id": 1463, "name": "culture clash"}, {"id":...  
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...  
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...  
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...  
4     [{"id": 818, "name": "based on novel"}, {"id":...  
...                                                 ...  
4804  [{"id": 5616, "name": "united states\u2013mexi...  
4805                                        

## My next approach will be

### 1. Creating dataframe with 3 columns
    . id
    . title
    . tags
### 2. To create tag column we will merge all other column beside id and title and will result in an paragraph for each movies
(I have taken top 3 actors only from cast and director only from crew)

In [11]:
movies.isnull().sum()

title       0
cast        0
crew        0
overview    3
genres      0
id          0
keywords    0
dtype: int64

In [12]:
movies = movies.dropna()

In [13]:
movies.isnull().sum()

title       0
cast        0
crew        0
overview    0
genres      0
id          0
keywords    0
dtype: int64

In [14]:
movies.duplicated().sum()

0

## Let's preprocess some columns to get suitable values rather than dictionary

In [15]:
movies.iloc[10].genres

'[{"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 28, "name": "Action"}, {"id": 878, "name": "Science Fiction"}]'

## Creating a helper function for preprocessing

## Converting string of list to list

In [16]:
import ast

def converter(obj):
    # Check if obj is a string, if so, use ast.literal_eval to convert
    if isinstance(obj, str):
        obj = ast.literal_eval(obj)
    
    # Assuming obj is a list of dictionaries, extract 'name'
    L = []
    for i in obj:
        L.append(i['name'])
    return L

In [17]:
result = converter(movies.iloc[0].genres)
result

['Action', 'Adventure', 'Fantasy', 'Science Fiction']

In [18]:
movies.iloc[:,4] = movies['genres'].apply(converter)

In [19]:
movies.head(6)

title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   
5                              Spider-Man 3   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   
5  [{"cast_id": 30, "character": "Peter Parker / ...   

                                                crew  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...   
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...   
2  [{"credit_id": "54805967c3a36829b5002c41", "de...   
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...   
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   
5  [{"credit_id": "52fe4252c3a36847f80151a5", "de...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   
5  The seemingly invincible Spider-Man goes up ag...   

                                          genres      id  \
0  [Action, Adventure, Fantasy, Science Fiction]   19995   
1                   [Adventure, Fantasy, Action]     285   
2                     [Action, Adventure, Crime]  206647   
3               [Action, Crime, Drama, Thriller]   49026   
4           [Action, Adventure, Science Fiction]   49529   
5                   [Fantasy, Action, Adventure]     559   

                                            keywords  
0  [{"id": 1463, "name": "culture clash"}, {"id":...  
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...  
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...  
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...  
4  [{"id": 818, "name": "based on novel"}, {"id":...  
5  [{"id": 851, "name": "dual identity"}, {"id": ...

In [20]:
movies.iloc[0, :].keywords

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [21]:
movies.iloc[0].keywords

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [22]:
result = converter(movies.iloc[0].keywords)
result

['culture clash',
 'future',
 'space war',
 'space colony',
 'society',
 'space travel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alien planet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'love affair',
 'anti war',
 'power relations',
 'mind and soul',
 '3d']

In [23]:
movies.iloc[:, -1] = movies['keywords'].apply(converter)

In [24]:
movies

title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4804                               El Mariachi   
4805                                 Newlyweds   
4806                 Signed, Sealed, Delivered   
4807                          Shanghai Calling   
4808                         My Date with Drew   

                                                   cast  \
0     [{"cast_id": 242, "character": "Jake Sully", "...   
1     [{"cast_id": 4, "character": "Captain Jack Spa...   
2     [{"cast_id": 1, "character": "James Bond", "cr...   
3     [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4     [{"cast_id": 5, "character": "John Carter", "c...   
...                                                 ...   
4804  [{"cast_id": 1, "character": "El Mariachi", "c...   
4805  [{"cast_id": 1, "character": "Buzzy", "credit_...   
4806  [{"cast_id": 8, "character": "Oliver O\u2019To...   
4807  [{"cast_id": 3, "character": "Sam", "credit_id...   
4808  [{"cast_id": 3, "character": "Herself", "credi...   

                                                   crew  \
0     [{"credit_id": "52fe48009251416c750aca23", "de...   
1     [{"credit_id": "52fe4232c3a36847f800b579", "de...   
2     [{"credit_id": "54805967c3a36829b5002c41", "de...   
3     [{"credit_id": "52fe4781c3a36847f81398c3", "de...   
4     [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   
...                                                 ...   
4804  [{"credit_id": "52fe44eec3a36847f80b280b", "de...   
4805  [{"credit_id": "52fe487dc3a368484e0fb013", "de...   
4806  [{"credit_id": "52fe4df3c3a36847f8275ecf", "de...   
4807  [{"credit_id": "52fe4ad9c3a368484e16a36b", "de...   
4808  [{"credit_id": "58ce021b9251415a390165d9", "de...   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4804  El Mariachi just wants to play his guitar and ...   
4805  A newlywed couple's honeymoon is upended by th...   
4806  "Signed, Sealed, Delivered" introduces a dedic...   
4807  When ambitious New York attorney Sam is sent t...   
4808  Ever since the second grade when he first saw ...   

                                             genres      id  \
0     [Action, Adventure, Fantasy, Science Fiction]   19995   
1                      [Adventure, Fantasy, Action]     285   
2                        [Action, Adventure, Crime]  206647   
3                  [Action, Crime, Drama, Thriller]   49026   
4              [Action, Adventure, Science Fiction]   49529   
...                                             ...     ...   
4804                      [Action, Crime, Thriller]    9367   
4805                              [Comedy, Romance]   72766   
4806             [Comedy, Drama, Romance, TV Movie]  231617   
4807                                             []  126186   
4808                                  [Documentary]   25975   

                                               keywords  
0     [culture clash, future, space war, space colon...  
1     [ocean, drug abuse, exotic island, east india ...  
2     [spy, based on novel, secret agent, sequel, mi...  
3     [dc comics, crime fighter, terrorist, secret i...  
4     [based on novel, mars, medallion, space travel...  
...                                                 ...  
4804  [united states–mexico barrier, legs, arms, pap...  
4805                                                 []  
4806  [date, love at first sight, 

In [25]:
movies.iloc[0].cast

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [26]:
result = converter(movies.iloc[0].cast)
result

['Sam Worthington',
 'Zoe Saldana',
 'Sigourney Weaver',
 'Stephen Lang',
 'Michelle Rodriguez',
 'Giovanni Ribisi',
 'Joel David Moore',
 'CCH Pounder',
 'Wes Studi',
 'Laz Alonso',
 'Dileep Rao',
 'Matt Gerald',
 'Sean Anthony Moran',
 'Jason Whyte',
 'Scott Lawrence',
 'Kelly Kilgour',
 'James Patrick Pitt',
 'Sean Patrick Murphy',
 'Peter Dillon',
 'Kevin Dorman',
 'Kelson Henderson',
 'David Van Horn',
 'Jacob Tomuri',
 'Michael Blain-Rozgay',
 'Jon Curry',
 'Luke Hawker',
 'Woody Schultz',
 'Peter Mensah',
 'Sonia Yee',
 'Jahnel Curfman',
 'Ilram Choi',
 'Kyla Warren',
 'Lisa Roumain',
 'Debra Wilson',
 'Chris Mala',
 'Taylor Kibby',
 'Jodie Landau',
 'Julie Lamm',
 'Cullen B. Madden',
 'Joseph Brady Madden',
 'Frankie Torres',
 'Austin Wilson',
 'Sara Wilson',
 'Tamica Washington-Miller',
 'Lucy Briant',
 'Nathan Meister',
 'Gerry Blair',
 'Matthew Chamberlain',
 'Paul Yates',
 'Wray Wilson',
 'James Gaylyn',
 'Melvin Leno Clark III',
 'Carvon Futrell',
 'Brandon Jelkes',
 'Mica

In [27]:
def cast_converter(obj):
    # Check if obj is a string, if so, use ast.literal_eval to convert
    if isinstance(obj, str):
        obj = ast.literal_eval(obj)
    
    # Assuming obj is a list of dictionaries, extract 'name'
    L = []
    val = 0
    for i in obj:
        val = val + 1
        L.append(i['name'])
        if val>=3:
            break
    return L

In [28]:
cast_converter(movies.iloc[0].cast)

['Sam Worthington', 'Zoe Saldana', 'Sigourney Weaver']

In [29]:
movies.iloc[:, 1] = movies['cast'].apply(cast_converter)

In [30]:
movies

title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4804                               El Mariachi   
4805                                 Newlyweds   
4806                 Signed, Sealed, Delivered   
4807                          Shanghai Calling   
4808                         My Date with Drew   

                                                   cast  \
0      [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1         [Johnny Depp, Orlando Bloom, Keira Knightley]   
2          [Daniel Craig, Christoph Waltz, Léa Seydoux]   
3          [Christian Bale, Michael Caine, Gary Oldman]   
4        [Taylor Kitsch, Lynn Collins, Samantha Morton]   
...                                                 ...   
4804  [Carlos Gallardo, Jaime de Hoyos, Peter Marqua...   
4805       [Edward Burns, Kerry Bishé, Marsha Dietlein]   
4806         [Eric Mabius, Kristin Booth, Crystal Lowe]   
4807          [Daniel Henney, Eliza Coupe, Bill Paxton]   
4808  [Drew Barrymore, Brian Herzlinger, Corey Feldman]   

                                                   crew  \
0     [{"credit_id": "52fe48009251416c750aca23", "de...   
1     [{"credit_id": "52fe4232c3a36847f800b579", "de...   
2     [{"credit_id": "54805967c3a36829b5002c41", "de...   
3     [{"credit_id": "52fe4781c3a36847f81398c3", "de...   
4     [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   
...                                                 ...   
4804  [{"credit_id": "52fe44eec3a36847f80b280b", "de...   
4805  [{"credit_id": "52fe487dc3a368484e0fb013", "de...   
4806  [{"credit_id": "52fe4df3c3a36847f8275ecf", "de...   
4807  [{"credit_id": "52fe4ad9c3a368484e16a36b", "de...   
4808  [{"credit_id": "58ce021b9251415a390165d9", "de...   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4804  El Mariachi just wants to play his guitar and ...   
4805  A newlywed couple's honeymoon is upended by th...   
4806  "Signed, Sealed, Delivered" introduces a dedic...   
4807  When ambitious New York attorney Sam is sent t...   
4808  Ever since the second grade when he first saw ...   

                                             genres      id  \
0     [Action, Adventure, Fantasy, Science Fiction]   19995   
1                      [Adventure, Fantasy, Action]     285   
2                        [Action, Adventure, Crime]  206647   
3                  [Action, Crime, Drama, Thriller]   49026   
4              [Action, Adventure, Science Fiction]   49529   
...                                             ...     ...   
4804                      [Action, Crime, Thriller]    9367   
4805                              [Comedy, Romance]   72766   
4806             [Comedy, Drama, Romance, TV Movie]  231617   
4807                                             []  126186   
4808                                  [Documentary]   25975   

                                               keywords  
0     [culture clash, future, space war, space colon...  
1     [ocean, drug abuse, exotic island, east india ...  
2     [spy, based on novel, secret agent, sequel, mi...  
3     [dc comics, crime fighter, terrorist, secret i...  
4     [based on novel, mars, medallion, space travel...  
...                                                 ...  
4804  [united states–mexico barrier, legs, arms, pap...  
4805                                                 []  
4806  [date, love at first sight, 

In [31]:
movies.iloc[1].crew

'[{"credit_id": "52fe4232c3a36847f800b579", "department": "Camera", "gender": 2, "id": 120, "job": "Director of Photography", "name": "Dariusz Wolski"}, {"credit_id": "52fe4232c3a36847f800b4fd", "department": "Directing", "gender": 2, "id": 1704, "job": "Director", "name": "Gore Verbinski"}, {"credit_id": "52fe4232c3a36847f800b54f", "department": "Production", "gender": 2, "id": 770, "job": "Producer", "name": "Jerry Bruckheimer"}, {"credit_id": "52fe4232c3a36847f800b503", "department": "Writing", "gender": 2, "id": 1705, "job": "Screenplay", "name": "Ted Elliott"}, {"credit_id": "52fe4232c3a36847f800b509", "department": "Writing", "gender": 2, "id": 1706, "job": "Screenplay", "name": "Terry Rossio"}, {"credit_id": "52fe4232c3a36847f800b57f", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "52fe4232c3a36847f800b585", "department": "Editing", "gender": 2, "id": 1722, "job": "Editor", "name": "Craig Wood"}, {"credit_id": "52f

In [32]:
def crew_converter(obj):
    # Check if obj is a string, if so, use ast.literal_eval to convert
    if isinstance(obj, str):
        obj = ast.literal_eval(obj)
    
    # Assuming obj is a list of dictionaries, extract 'name'
    L = []
    for i in obj:
        if i['job'] == 'Director':
            L.append(i['name'])
            break
    return L

In [33]:
crew_converter(movies.iloc[0].crew)

['James Cameron']

In [34]:
movies.iloc[:, 2] = movies['crew'].apply(crew_converter)

In [35]:
movies

title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4804                               El Mariachi   
4805                                 Newlyweds   
4806                 Signed, Sealed, Delivered   
4807                          Shanghai Calling   
4808                         My Date with Drew   

                                                   cast                 crew  \
0      [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]   
1         [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]   
2          [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]   
3          [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]   
4        [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]   
...                                                 ...                  ...   
4804  [Carlos Gallardo, Jaime de Hoyos, Peter Marqua...   [Robert Rodriguez]   
4805       [Edward Burns, Kerry Bishé, Marsha Dietlein]       [Edward Burns]   
4806         [Eric Mabius, Kristin Booth, Crystal Lowe]        [Scott Smith]   
4807          [Daniel Henney, Eliza Coupe, Bill Paxton]        [Daniel Hsia]   
4808  [Drew Barrymore, Brian Herzlinger, Corey Feldman]   [Brian Herzlinger]   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4804  El Mariachi just wants to play his guitar and ...   
4805  A newlywed couple's honeymoon is upended by th...   
4806  "Signed, Sealed, Delivered" introduces a dedic...   
4807  When ambitious New York attorney Sam is sent t...   
4808  Ever since the second grade when he first saw ...   

                                             genres      id  \
0     [Action, Adventure, Fantasy, Science Fiction]   19995   
1                      [Adventure, Fantasy, Action]     285   
2                        [Action, Adventure, Crime]  206647   
3                  [Action, Crime, Drama, Thriller]   49026   
4              [Action, Adventure, Science Fiction]   49529   
...                                             ...     ...   
4804                      [Action, Crime, Thriller]    9367   
4805                              [Comedy, Romance]   72766   
4806             [Comedy, Drama, Romance, TV Movie]  231617   
4807                                             []  126186   
4808                                  [Documentary]   25975   

                                               keywords  
0     [culture clash, future, space war, space colon...  
1     [ocean, drug abuse, exotic island, east india ...  
2     [spy, based on novel, secret agent, sequel, mi...  
3     [dc comics, crime fighter, terrorist, secret i...  
4     [based on novel, mars, medallion, space travel...  
...                                                 ...  
4804  [united states–mexico barrier, legs, arms, pap...  
4805                                                 []  
4806  [date, love at first sight, narration, investi...  
4807                                                 []  
4808          [obsession, camcorder, crush, dream girl]  

[4806 rows x 7 columns]

## Let's convert overview into a list so it will be easier to merge with other columns

In [36]:
movies['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [37]:
movies.iloc[:, 3] = movies['overview'].apply(lambda x: x.split())

In [38]:
movies

title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4804                               El Mariachi   
4805                                 Newlyweds   
4806                 Signed, Sealed, Delivered   
4807                          Shanghai Calling   
4808                         My Date with Drew   

                                                   cast                 crew  \
0      [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]   
1         [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]   
2          [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]   
3          [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]   
4        [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]   
...                                                 ...                  ...   
4804  [Carlos Gallardo, Jaime de Hoyos, Peter Marqua...   [Robert Rodriguez]   
4805       [Edward Burns, Kerry Bishé, Marsha Dietlein]       [Edward Burns]   
4806         [Eric Mabius, Kristin Booth, Crystal Lowe]        [Scott Smith]   
4807          [Daniel Henney, Eliza Coupe, Bill Paxton]        [Daniel Hsia]   
4808  [Drew Barrymore, Brian Herzlinger, Corey Feldman]   [Brian Herzlinger]   

                                               overview  \
0     [In, the, 22nd, century,, a, paraplegic, Marin...   
1     [Captain, Barbossa,, long, believed, to, be, d...   
2     [A, cryptic, message, from, Bond’s, past, send...   
3     [Following, the, death, of, District, Attorney...   
4     [John, Carter, is, a, war-weary,, former, mili...   
...                                                 ...   
4804  [El, Mariachi, just, wants, to, play, his, gui...   
4805  [A, newlywed, couple's, honeymoon, is, upended...   
4806  ["Signed,, Sealed,, Delivered", introduces, a,...   
4807  [When, ambitious, New, York, attorney, Sam, is...   
4808  [Ever, since, the, second, grade, when, he, fi...   

                                             genres      id  \
0     [Action, Adventure, Fantasy, Science Fiction]   19995   
1                      [Adventure, Fantasy, Action]     285   
2                        [Action, Adventure, Crime]  206647   
3                  [Action, Crime, Drama, Thriller]   49026   
4              [Action, Adventure, Science Fiction]   49529   
...                                             ...     ...   
4804                      [Action, Crime, Thriller]    9367   
4805                              [Comedy, Romance]   72766   
4806             [Comedy, Drama, Romance, TV Movie]  231617   
4807                                             []  126186   
4808                                  [Documentary]   25975   

                                               keywords  
0     [culture clash, future, space war, space colon...  
1     [ocean, drug abuse, exotic island, east india ...  
2     [spy, based on novel, secret agent, sequel, mi...  
3     [dc comics, crime fighter, terrorist, secret i...  
4     [based on novel, mars, medallion, space travel...  
...                                                 ...  
4804  [united states–mexico barrier, legs, arms, pap...  
4805                                                 []  
4806  [date, love at first sight, narration, investi...  
4807                                                 []  
4808          [obsession, camcorder, crush, dream girl]  

[4806 rows x 7 columns]

## Removing spaces between works, it is very important for classification

In [39]:
movies.iloc[:, 4] = movies['genres'].apply(lambda x: [i.replace(" ", '') for i in x])
movies.iloc[:, -1] = movies['keywords'].apply(lambda x: [i.replace(" ", '') for i in x])
movies.iloc[:, 1] = movies['cast'].apply(lambda x: [i.replace(" ", '') for i in x])
movies.iloc[:, 2] = movies['crew'].apply(lambda x: [i.replace(" ", '') for i in x])

In [40]:
movies

title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4804                               El Mariachi   
4805                                 Newlyweds   
4806                 Signed, Sealed, Delivered   
4807                          Shanghai Calling   
4808                         My Date with Drew   

                                                cast                crew  \
0      [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1         [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2          [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3          [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4        [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   
...                                              ...                 ...   
4804  [CarlosGallardo, JaimedeHoyos, PeterMarquardt]   [RobertRodriguez]   
4805       [EdwardBurns, KerryBishé, MarshaDietlein]       [EdwardBurns]   
4806         [EricMabius, KristinBooth, CrystalLowe]        [ScottSmith]   
4807          [DanielHenney, ElizaCoupe, BillPaxton]        [DanielHsia]   
4808  [DrewBarrymore, BrianHerzlinger, CoreyFeldman]   [BrianHerzlinger]   

                                               overview  \
0     [In, the, 22nd, century,, a, paraplegic, Marin...   
1     [Captain, Barbossa,, long, believed, to, be, d...   
2     [A, cryptic, message, from, Bond’s, past, send...   
3     [Following, the, death, of, District, Attorney...   
4     [John, Carter, is, a, war-weary,, former, mili...   
...                                                 ...   
4804  [El, Mariachi, just, wants, to, play, his, gui...   
4805  [A, newlywed, couple's, honeymoon, is, upended...   
4806  ["Signed,, Sealed,, Delivered", introduces, a,...   
4807  [When, ambitious, New, York, attorney, Sam, is...   
4808  [Ever, since, the, second, grade, when, he, fi...   

                                            genres      id  \
0     [Action, Adventure, Fantasy, ScienceFiction]   19995   
1                     [Adventure, Fantasy, Action]     285   
2                       [Action, Adventure, Crime]  206647   
3                 [Action, Crime, Drama, Thriller]   49026   
4              [Action, Adventure, ScienceFiction]   49529   
...                                            ...     ...   
4804                     [Action, Crime, Thriller]    9367   
4805                             [Comedy, Romance]   72766   
4806             [Comedy, Drama, Romance, TVMovie]  231617   
4807                                            []  126186   
4808                                 [Documentary]   25975   

                                               keywords  
0     [cultureclash, future, spacewar, spacecolony, ...  
1     [ocean, drugabuse, exoticisland, eastindiatrad...  
2     [spy, basedonnovel, secretagent, sequel, mi6, ...  
3     [dccomics, crimefighter, terrorist, secretiden...  
4     [basedonnovel, mars, medallion, spacetravel, p...  
...                                                 ...  
4804  [unitedstates–mexicobarrier, legs, arms, paper...  
4805                                                 []  
4806  [date, loveatfirstsight, narration, investigat...  
4807                                                 []  
4808           [obsession, camcorder, crush, dreamgirl]  

[4806 rows x 7 columns]

## Now merging 5 columns in a tag column

In [41]:
movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

C:\Users\acer\AppData\Local\Temp\ipykernel_1480\603755214.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']


In [42]:
movies

title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4804                               El Mariachi   
4805                                 Newlyweds   
4806                 Signed, Sealed, Delivered   
4807                          Shanghai Calling   
4808                         My Date with Drew   

                                                cast                crew  \
0      [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1         [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2          [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3          [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4        [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   
...                                              ...                 ...   
4804  [CarlosGallardo, JaimedeHoyos, PeterMarquardt]   [RobertRodriguez]   
4805       [EdwardBurns, KerryBishé, MarshaDietlein]       [EdwardBurns]   
4806         [EricMabius, KristinBooth, CrystalLowe]        [ScottSmith]   
4807          [DanielHenney, ElizaCoupe, BillPaxton]        [DanielHsia]   
4808  [DrewBarrymore, BrianHerzlinger, CoreyFeldman]   [BrianHerzlinger]   

                                               overview  \
0     [In, the, 22nd, century,, a, paraplegic, Marin...   
1     [Captain, Barbossa,, long, believed, to, be, d...   
2     [A, cryptic, message, from, Bond’s, past, send...   
3     [Following, the, death, of, District, Attorney...   
4     [John, Carter, is, a, war-weary,, former, mili...   
...                                                 ...   
4804  [El, Mariachi, just, wants, to, play, his, gui...   
4805  [A, newlywed, couple's, honeymoon, is, upended...   
4806  ["Signed,, Sealed,, Delivered", introduces, a,...   
4807  [When, ambitious, New, York, attorney, Sam, is...   
4808  [Ever, since, the, second, grade, when, he, fi...   

                                            genres      id  \
0     [Action, Adventure, Fantasy, ScienceFiction]   19995   
1                     [Adventure, Fantasy, Action]     285   
2                       [Action, Adventure, Crime]  206647   
3                 [Action, Crime, Drama, Thriller]   49026   
4              [Action, Adventure, ScienceFiction]   49529   
...                                            ...     ...   
4804                     [Action, Crime, Thriller]    9367   
4805                             [Comedy, Romance]   72766   
4806             [Comedy, Drama, Romance, TVMovie]  231617   
4807                                            []  126186   
4808                                 [Documentary]   25975   

                                               keywords  \
0     [cultureclash, future, spacewar, spacecolony, ...   
1     [ocean, drugabuse, exoticisland, eastindiatrad...   
2     [spy, basedonnovel, secretagent, sequel, mi6, ...   
3     [dccomics, crimefighter, terrorist, secretiden...   
4     [basedonnovel, mars, medallion, spacetravel, p...   
...                                                 ...   
4804  [unitedstates–mexicobarrier, legs, arms, paper...   
4805                                                 []   
4806  [date, loveatfirstsight, narration, investigat...   
4807                                                 []   
4808           [obsession, camcorder, crush, dreamgirl]   

                                                   tags  
0     [In, the, 22nd, century,, a, paraplegic, Marin...  
1     [Captain, Barbossa,, long, believed, to, be, d...  
2     [A, cryptic, message, from, Bond’s, past, send...  
3     [Following, the, death, of, District, Attorney...  
4     [John, Carter, is, a, war-weary,, former, mili...  
...             

In [43]:
final_df = movies[["id", "title", "tags"]]

In [44]:
final_df

id                                     title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4804    9367                               El Mariachi   
4805   72766                                 Newlyweds   
4806  231617                 Signed, Sealed, Delivered   
4807  126186                          Shanghai Calling   
4808   25975                         My Date with Drew   

                                                   tags  
0     [In, the, 22nd, century,, a, paraplegic, Marin...  
1     [Captain, Barbossa,, long, believed, to, be, d...  
2     [A, cryptic, message, from, Bond’s, past, send...  
3     [Following, the, death, of, District, Attorney...  
4     [John, Carter, is, a, war-weary,, former, mili...  
...                                                 ...  
4804  [El, Mariachi, just, wants, to, play, his, gui...  
4805  [A, newlywed, couple's, honeymoon, is, upended...  
4806  ["Signed,, Sealed,, Delivered", introduces, a,...  
4807  [When, ambitious, New, York, attorney, Sam, is...  
4808  [Ever, since, the, second, grade, when, he, fi...  

[4806 rows x 3 columns]

In [45]:
print(final_df['tags'].apply(lambda x: isinstance(x, list) and all(isinstance(i, str) for i in x)).head())
#  checking if each row contains list of strings or not

0    True
1    True
2    True
3    True
4    True
Name: tags, dtype: bool


In [46]:
final_df.iloc[:, -1] = final_df['tags'].apply(lambda x: " ".join(x))
final_df

id                                     title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4804    9367                               El Mariachi   
4805   72766                                 Newlyweds   
4806  231617                 Signed, Sealed, Delivered   
4807  126186                          Shanghai Calling   
4808   25975                         My Date with Drew   

                                                   tags  
0     In the 22nd century, a paraplegic Marine is di...  
1     Captain Barbossa, long believed to be dead, ha...  
2     A cryptic message from Bond’s past sends him o...  
3     Following the death of District Attorney Harve...  
4     John Carter is a war-weary, former military ca...  
...                                                 ...  
4804  El Mariachi just wants to play his guitar and ...  
4805  A newlywed couple's honeymoon is upended by th...  
4806  "Signed, Sealed, Delivered" introduces a dedic...  
4807  When ambitious New York attorney Sam is sent t...  
4808  Ever since the second grade when he first saw ...  

[4806 rows x 3 columns]

In [47]:
final_df['tags'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

In [48]:
# converting to lower case
final_df.iloc[:, -1] = final_df['tags'].apply(lambda x: x.lower())
final_df

id                                     title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4804    9367                               El Mariachi   
4805   72766                                 Newlyweds   
4806  231617                 Signed, Sealed, Delivered   
4807  126186                          Shanghai Calling   
4808   25975                         My Date with Drew   

                                                   tags  
0     in the 22nd century, a paraplegic marine is di...  
1     captain barbossa, long believed to be dead, ha...  
2     a cryptic message from bond’s past sends him o...  
3     following the death of district attorney harve...  
4     john carter is a war-weary, former military ca...  
...                                                 ...  
4804  el mariachi just wants to play his guitar and ...  
4805  a newlywed couple's honeymoon is upended by th...  
4806  "signed, sealed, delivered" introduces a dedic...  
4807  when ambitious new york attorney sam is sent t...  
4808  ever since the second grade when he first saw ...  

[4806 rows x 3 columns]

## These types of issues exists in the feature names


    59                         accident
    60                       accidental
    61                     accidentally
    
    We know they represent the same thing

## Therefore stemming is applied to solve this issues before vectorizing to get feature names

In [49]:
!pip install nltk


[notice] A new release of pip is available: 24.2 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [50]:
import nltk

In [51]:
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [52]:
def stem(text):
    y = []
    for i in text.split():
        y.append(ps.stem(i))
    return " ".join(y)

In [53]:
final_df.iloc[:, -1]= final_df['tags'].apply(stem)

In [54]:
final_df.iloc[:, 1] = final_df['title'].str.lower()
final_df

id                                     title  \
0      19995                                    avatar   
1        285  pirates of the caribbean: at world's end   
2     206647                                   spectre   
3      49026                     the dark knight rises   
4      49529                               john carter   
...      ...                                       ...   
4804    9367                               el mariachi   
4805   72766                                 newlyweds   
4806  231617                 signed, sealed, delivered   
4807  126186                          shanghai calling   
4808   25975                         my date with drew   

                                                   tags  
0     in the 22nd century, a parapleg marin is dispa...  
1     captain barbossa, long believ to be dead, ha c...  
2     a cryptic messag from bond’ past send him on a...  
3     follow the death of district attorney harvey d...  
4     john carter is a war-weary, former militari ca...  
...                                                 ...  
4804  el mariachi just want to play hi guitar and ca...  
4805  a newlyw couple' honeymoon is upend by the arr...  
4806  "signed, sealed, delivered" introduc a dedic q...  
4807  when ambiti new york attorney sam is sent to s...  
4808  ever sinc the second grade when he first saw h...  

[4806 rows x 3 columns]

In [55]:
final_df.to_csv("processed_movies.csv")

## Text Vectorization for similarity calculation

For recommendation we suggest movies based on their similarity, and similarity is calculated based on generated tags.
For this we need to vectorize the tags first and after converting them into a vector we can perform calculation to find similarity between any 2 movies.
Based on this we will suggest the movie to the user.

## Different techniques for text vectorization
    
1. Bag of words (I am using this one)
2. TF IDF
3. word2vec

## Working of bag of words

The Bag of Words (BoW) model is a simple way to convert text into numerical data for machine learning. Here's how it works in simple steps:

  ### 1. Combing words
  All words in the entire dataset is merge into a bag of words
  
  ### 2. calculating frequency
  Calculate frequency for each word in the bag of words for every movie
  
  ### 3. Vectorization
The frequency for each word is considered a vector for that movie  

### 4. Ignoring stop words
All stops words like but, in, a, the etc are not considered for this frequency calculation because they have no meaning in calculating similarity between movies.
  

In [56]:
from sklearn.feature_extraction.text import CountVectorizer

In [57]:
CV = CountVectorizer(max_features = 5000, stop_words = 'english')

In [58]:
CV.fit_transform(final_df['tags']).toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [59]:
vectors = CV.fit_transform(final_df['tags']).toarray()

In [60]:
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [61]:
feature_names = CV.get_feature_names_out()

In [62]:
# Convert to pandas Series to display without truncation
pd.set_option('display.max_rows', None)

# Display the feature names
print(pd.Series(feature_names))

0                              000
1                              007
2                               10
3                              100
4                               11
5                               12
6                               13
7                               14
8                               15
9                               16
10                              17
11                            17th
12                              18
13                            18th
14                     18thcenturi
15                              19
16                            1910
17                            1920
18                            1930
19                            1940
20                            1944
21                            1950
22                           1950s
23                            1960
24                           1960s
25                            1970
26                           1970s
27                            1971
28                  

## Calculating similarity through cosine similarity

I have calculated cosine similarity between every vectors to mesaure its similarity with one another

In [63]:
from sklearn.metrics.pairwise import cosine_similarity

In [64]:
similarity = cosine_similarity(vectors)

In [74]:
import pickle

# Save the similarity matrix to a file
with open("similarity.pkl", "wb") as file:
    pickle.dump(similarity, file)
print("Similarity matrix saved successfully!")


Similarity matrix saved successfully!


In [65]:
similarity[0]

array([1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
       0.        ])

In [66]:
final_df[final_df['title'] == 'batman'].index[0]

1360

In [67]:
similarity[0]

array([1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
       0.        ])

In [75]:
# Load the similarity matrix from the file
with open("similarity.pkl", "rb") as file:
    similarity = pickle.load(file)
print("Similarity matrix loaded successfully!")


Similarity matrix loaded successfully!


In [76]:
def recommend(movie):
    # Converting the input movie title to lowercase
    movie = movie.lower()

    # Check if the movie exists in the DataFrame
    if movie not in final_df['title'].values:
        print(f"'{movie.title()}' not found in the database.")
        return

    # Find the index of the movie in the final DataFrame
    index = final_df[final_df['title'] == movie].index[0]
    
    # Get the similarity scores of this movie with all other movies
    similarity_with_other_movies = similarity[index]
    
    # Sort the similarity scores in descending order and get the top 5 most similar movies (ignoring the movie itself)
    similar_movies = sorted(list(enumerate(similarity_with_other_movies)), key=lambda x: x[1], reverse=True)[1:6]
    
    # Print the titles of the recommended movies in capatalized format
    for i in similar_movies:
        print(final_df.iloc[i[0]]['title'].title())


## Explanation of Each Line:

`def recommend(movie):`

Defines a function `recommend` that takes one argument, `movie`, which is the title of the movie you want to find recommendations for.

---

`index = final_df[final_df['title'] == movie].index[0]`

- This line searches for the row in `final_df` where the 'title' column matches the given movie title.
- It retrieves the index of that row using `.index[0]`, which is the index of the first movie found.
- Example: If `movie = 'Avatar'`, and 'Avatar' exists at index 1 in `final_df`, this line will return `1`.

---

`similarity_with_other_movies = similarity[index]`

- `similarity` is assumed to be a precomputed similarity matrix (e.g., cosine similarity).
- This line retrieves the similarity scores between the given movie and all other movies using the index we found earlier.
- The result is a list or array where each element represents how similar another movie is to the one at the index position.

---

`similar_movies = sorted(list(enumerate(similarity_with_other_movies)), key=lambda x: x[1], reverse=True)[1:6]`

- `enumerate(similarity_with_other_movies)`: This creates a list of tuples, where each tuple is of the form `(index, similarity_score)` for every movie in the similarity matrix.
- `sorted(..., key=lambda x: x[1], reverse=True)`: This sorts the list of tuples by similarity scores (`x[1]`), in descending order (highest similarity first).
- `[1:6]`: This slices the sorted list to get the 5 most similar movies, skipping the first one (`[1]`), because the highest similarity is the movie itself (similarity with itself is 1).
- **Result**: You now have a list of the 5 most similar movies (their indices and similarity scores).

---

`for i in similar_movies:`

- Iterates over the list of similar movies. Each `i` is a tuple containing the index of a movie and its similarity score (e.g., `(3, 0.85)`).

---

`print(final_df.iloc[i[0]]['title'])`

- For each similar movie, it retrieves the movie title using `.iloc[i[0]]`, where `i[0]` is the index of the similar movie.
- `final_df.iloc[i[0]]['title']`: This looks up the movie title at the index `i[0]` and prints it.


In [78]:
recommend("titanic")

The Notebook
Under The Same Moon
Ghost Ship
The Bounty
Pirates Of The Caribbean: On Stranger Tides
